### ISSUE TITLE
Compilation error with iszero with glibc (>2.24)

### ISSUE THREAD

**Issue by fdmalone** [QMCPACK/qmcpack#429]
I ran into the following when trying to compile qmcpack on my laptop (linux, gcc v7.2.0):

```
[ 15%] Building CXX object src/CMakeFiles/qmcbase.dir/Lattice/Uniform3DGridLayout.cpp.o
[ 15%] Building CXX object src/CMakeFiles/qmcbase.dir/Particle/ParticleSet.cpp.o
[ 15%] Building CXX object src/CMakeFiles/qmcbase.dir/Particle/ParticleSet.BC.cpp.o
[ 16%] Building CXX object src/CMakeFiles/qmcbase.dir/Particle/MCWalkerConfiguration.cpp.o
In file included from /home/fionn/projects/qmcpack_public/src/Estimators/accumulators.h:27:0,
                 from /home/fionn/projects/qmcpack_public/src/Estimators/ScalarEstimatorBase.h:24,
                 from /home/fionn/projects/qmcpack_public/src/Estimators/EstimatorManagerBase.h:30,
                 from /home/fionn/projects/qmcpack_public/src/QMCDrivers/QMCDriver.h:32,
                 from /home/fionn/projects/qmcpack_public/src/Particle/MCWalkerConfiguration.cpp:33:
/home/fionn/projects/qmcpack_public/src/config/stdlib/limits.h:17:13: error: redefinition of ‘template<class T> bool iszero(T)’
 inline bool iszero(T a)
             ^~~~~~
In file included from /usr/include/c++/7.2.0/cmath:45:0,
                 from /usr/include/c++/7.2.0/complex:44,
                 from /home/fionn/projects/qmcpack_public/src/Configuration.h:24,
                 from /home/fionn/projects/qmcpack_public/src/Particle/ParticleSet.h:25,
                 from /home/fionn/projects/qmcpack_public/src/Particle/MCWalkerConfiguration.h:28,
                 from /home/fionn/projects/qmcpack_public/src/Particle/MCWalkerConfiguration.cpp:24:
/usr/include/math.h:551:1: note: ‘template<class __T> bool iszero(__T)’ previously declared here
 iszero (__T __val)
 ^~~~~~
make[2]: *** [src/CMakeFiles/qmcbase.dir/build.make:135: src/CMakeFiles/qmcbase.dir/Particle/MCWalkerConfiguration.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1237: src/CMakeFiles/qmcbase.dir/all] Error 2
make: *** [Makefile:95: all] Error 2
```

It looks like iszero was added to math.h from v2.25 of glibc onwards defined as:
```
template <class __T> inline bool
iszero (__T __val)
{
  return __val == 0;
}
```

Is there some way around this besided downgrading glibc? If I comment out these lines from math.h (!) the code compiles.

**Comment by prckent** [comment]
@markdewing Are you able to check this? We would have to setup a VM here to check (something we should have available in future)

**Comment by prckent** [comment]
@fdmalone Which Linux are you using?

**Comment by markdewing** [comment]
I'lll find a way to check this.   In the worst case, I'll create a VM with a newer linux distribution.

**Comment by ye-luo** [comment]
@markdewing You can use the following to guard codes as I did for the allocator.
#if __GLIBC__ == 2 && __GLIBC_MINOR__ >= 25

**Comment by fdmalone** [comment]
I'm using arch linux 4.13.7-1

**Comment by prckent** [comment]
This is something we might want cmake to check

**Comment by ye-luo** [comment]
We need both.
For the allocator, by guarding codes, we don't need to upset users in CMake.
However, my recent investigation of libmvec shows that gcc+glibc2.23 is bad. This has to be handled by CMake.

**Comment by Paul-St-Young** [comment]
@fdmalone I have seen the exact same error message while compiling on Fedora 26. The temporary fix is to comment out the contents of `src/config/stdlib/limits.h`.

**Comment by fdmalone** [comment]
@Paul-St-Young It seems as if the qmcpack iszero function should take precedence over that provided by glibc since it does a safer comparison to zero, at least at a glance.

**Comment by prckent** [comment]
ISO TS 18661-1 defines iszero in <math.h>. We need feature detection for this function and should use it appropriately when available.

**Comment by prckent** [comment]
We need to be alert to issues associated with newer glibc and gcc versions. e.g. The recently released Ubuntu 17.10 has glibc 2.26 and gcc 7.2

**Comment by grahamlopez** [comment]
This is on my list to explore/handle with both CI and cmake

**Comment by prckent** [comment]
@Paul-St-Young @fdmalone This should be fixed now due to Mark's #431 . Please can you confirm before I close this issue?

**Comment by Paul-St-Young** [comment]
@prckent I can confirm that the `iszero` error disappeared on:
- Fedora 24 having gcc 6.3.1 & glib 2.23
- Fedora 26 having gcc 7.1.1 & glib 2.25

**Comment by fdmalone** [comment]
@prckent I can also confirm that the issue is resolved for arch linux 4.13.7-1 with gcc 7.2.0 and glibc 2.26. Thanks!

**Comment by prckent** [comment]
Thanks both. I also can confirm Ubuntu 17.10 is OK.
